File: typescript/tests/integration/agent-bay-list.integration.test.ts
import { AgentBay, CreateSessionParams, Session } from "../../src";

function getTestAPIKey(): string {
  const apiKey = process.env.AGENTBAY_API_KEY;
  if (!apiKey) {
    console.warn(
      "Warning: AGENTBAY_API_KEY environment variable not set. Using default test key."
    );
    return "akm-xxx"; // Replace with your test API key
  }
  return apiKey;
}

function generateUniqueId(): string {
  const timestamp = Date.now() * 1000 + Math.floor(Math.random() * 1000);
  const randomPart = Math.floor(Math.random() * 10000);
  return `${timestamp}-${randomPart}`;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

describe("AgentBay.list() Integration Tests", () => {
  let agentBay: AgentBay;
  let uniqueId: string;
  let testSessions: Session[] = [];

  beforeAll(async () => {
    const apiKey = getTestAPIKey();
    agentBay = new AgentBay({ apiKey });
    uniqueId = generateUniqueId();

    console.log(`Using unique ID for test: ${uniqueId}`);

    // Create multiple sessions with different labels for testing
    console.log("Creating session 1 with dev environment...");
    const params1: CreateSessionParams = {
      labels: {
        project: `list-test-${uniqueId}`,
        environment: "dev",
        owner: `test-${uniqueId}`,
      },
    };
    const result1 = await agentBay.create(params1);
    if (result1.success && result1.session) {
      testSessions.push(result1.session);
      console.log(`Session 1 created: ${result1.session.sessionId}`);
      console.log(`Request ID: ${result1.requestId}`);
    }

    console.log("Creating session 2 with staging environment...");
    const params2: CreateSessionParams = {
      labels: {
        project: `list-test-${uniqueId}`,
        environment: "staging",
        owner: `test-${uniqueId}`,
      },
    };
    const result2 = await agentBay.create(params2);
    if (result2.success && result2.session) {
      testSessions.push(result2.session);
      console.log(`Session 2 created: ${result2.session.sessionId}`);
      console.log(`Request ID: ${result2.requestId}`);
    }

    console.log("Creating session 3 with prod environment...");
    const params3: CreateSessionParams = {
      labels: {
        project: `list-test-${uniqueId}`,
        environment: "prod",
        owner: `test-${uniqueId}`,
      },
    };
    const result3 = await agentBay.create(params3);
    if (result3.success && result3.session) {
      testSessions.push(result3.session);
      console.log(`Session 3 created: ${result3.session.sessionId}`);
      console.log(`Request ID: ${result3.requestId}`);
    }

    // Verify all sessions were created
    if (testSessions.length !== 3) {
      throw new Error(`Failed to create all 3 test sessions. Only created ${testSessions.length} sessions.`);
    }

    // Wait longer for sessions to be fully created and labels to propagate
    console.log("Waiting 5 seconds for labels to propagate...");
    await sleep(5000);
  }, 60000);

  afterAll(async () => {
    console.log("Cleaning up: Deleting all test sessions...");
    for (const session of testSessions) {
      try {
        const result = await agentBay.delete(session);
        console.log(
          `Session ${session.sessionId} deleted. Success: ${result.success}, Request ID: ${result.requestId}`
        );
      } catch (error) {
        console.warn(
          `Warning: Error deleting session ${session.sessionId}: ${error}`
        );
      }
    }
  }, 60000);

  test("should list all sessions without any label filter", async () => {
    console.log("\n=== Testing list() without labels ===");

    const result = await agentBay.list();

    // Verify the result
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
    expect(result.requestId).not.toBe("");
    expect(result.sessionIds).toBeDefined();

    console.log(`Total sessions found: ${result.totalCount}`);
    console.log(`Sessions in current page: ${result.sessionIds.length}`);
    console.log(`Request ID: ${result.requestId}`);
  });

  test("should list sessions with a single label filter", async () => {
    console.log("\n=== Testing list() with single label ===");

    const result = await agentBay.list({
      project: `list-test-${uniqueId}`,
    });

    // Verify the result
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
    expect(result.requestId).not.toBe("");
    expect(result.sessionIds.length).toBeGreaterThanOrEqual(3);

    // Verify all returned sessions have the expected label
    const sessionIds = testSessions.map((s) => s.sessionId);
    const foundCount = result.sessionIds.filter((sid) =>
      sessionIds.includes(sid)
    ).length;

    expect(foundCount).toBe(3);

    console.log(`Found ${foundCount} test sessions`);
    console.log(`Total sessions with label: ${result.sessionIds.length}`);
    console.log(`Request ID: ${result.requestId}`);
  });

  test("should list sessions with multiple label filters", async () => {
    console.log("\n=== Testing list() with multiple labels ===");

    const result = await agentBay.list({
      project: `list-test-${uniqueId}`,
      environment: "dev",
    });

    // Verify the result
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
    expect(result.requestId).not.toBe("");
    expect(result.sessionIds.length).toBeGreaterThanOrEqual(1);

    // Verify the dev session is in the results
    const devSessionId = testSessions[0].sessionId;
    const found = result.sessionIds.includes(devSessionId);

    expect(found).toBe(true);

    console.log(`Found dev session: ${found}`);
    console.log(`Total matching sessions: ${result.sessionIds.length}`);
    console.log(`Request ID: ${result.requestId}`);
  });

  test("should list sessions with pagination parameters", async () => {
    console.log("\n=== Testing list() with pagination ===");

    // List first page with limit of 2
    const resultPage1 = await agentBay.list(
      { project: `list-test-${uniqueId}` },
      1,
      2
    );

    // Verify first page
    expect(resultPage1.success).toBe(true);
    expect(resultPage1.requestId).toBeDefined();
    expect(resultPage1.requestId).not.toBe("");
    expect(resultPage1.data.length).toBeLessThanOrEqual(2);

    console.log(`Page 1 - Found ${resultPage1.data.length} sessions`);
    console.log(`Request ID: ${resultPage1.requestId}`);

    // If there are more results, test page 2
    if (resultPage1.nextToken) {
      const resultPage2 = await agentBay.list(
        { project: `list-test-${uniqueId}` },
        2,
        2
      );

      expect(resultPage2.success).toBe(true);
      expect(resultPage2.requestId).toBeDefined();
      expect(resultPage2.requestId).not.toBe("");

      console.log(`Page 2 - Found ${resultPage2.data.length} sessions`);
      console.log(`Request ID: ${resultPage2.requestId}`);
    }
  });

  test("should return empty results for non-matching labels", async () => {
    console.log("\n=== Testing list() with non-matching label ===");

    const result = await agentBay.list({
      project: `list-test-${uniqueId}`,
      environment: "nonexistent",
    });

    // Verify the result
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
    expect(result.requestId).not.toBe("");

    // Verify our test sessions are NOT in the results
    const sessionIds = testSessions.map((s) => s.sessionId);
    const foundCount = result.sessionIds.filter((sid) =>
      sessionIds.includes(sid)
    ).length;

    expect(foundCount).toBe(0);

    console.log(`Correctly found ${foundCount} test sessions (expected 0)`);
    console.log(`Request ID: ${result.requestId}`);
  });

  test("should use default limit of 10 when not specified", async () => {
    console.log("\n=== Testing list() with default limit ===");

    const result = await agentBay.list({ owner: `test-${uniqueId}` });

    // Verify the result
    expect(result.success).toBe(true);
    expect(result.requestId).toBeDefined();
    expect(result.requestId).not.toBe("");
    expect(result.maxResults).toBe(10);

    console.log(`Max results: ${result.maxResults}`);
    console.log(`Request ID: ${result.requestId}`);
  });

  test("should include request_id in all responses", async () => {
    console.log("\n=== Testing list() request_id presence ===");

    // Test 1: No labels
    const result1 = await agentBay.list();
    expect(result1.requestId).toBeDefined();
    expect(result1.requestId).not.toBe("");
    console.log(`Test 1 Request ID: ${result1.requestId}`);

    // Test 2: With labels
    const result2 = await agentBay.list({
      project: `list-test-${uniqueId}`,
    });
    expect(result2.requestId).toBeDefined();
    expect(result2.requestId).not.toBe("");
    console.log(`Test 2 Request ID: ${result2.requestId}`);

    // Test 3: With pagination
    const result3 = await agentBay.list(
      { project: `list-test-${uniqueId}` },
      1,
      5
    );
    expect(result3.requestId).toBeDefined();
    expect(result3.requestId).not.toBe("");
    console.log(`Test 3 Request ID: ${result3.requestId}`);
  });

  test("should handle error scenarios correctly", async () => {
    console.log("\\n=== Testing list() error scenarios ===");

    // Test 1: Invalid page number (page = 0)
    console.log("\\nTest 1: page=0 should return error");
    const result1 = await agentBay.list(
      { project: `list-test-${uniqueId}` },
      0,
      5
    );
    expect(result1.success).toBe(false);
    expect(result1.errorMessage).toBeDefined();
    expect(result1.errorMessage).toContain("Page number must be >= 1");
    console.log(`✓ Correctly rejected page=0: ${result1.errorMessage}`);

    // Test 2: Invalid page number (page = -1)
    console.log("\\nTest 2: page=-1 should return error");
    const result2 = await agentBay.list(
      { project: `list-test-${uniqueId}` },
      -1,
      5
    );
    expect(result2.success).toBe(false);
    expect(result2.errorMessage).toBeDefined();
    expect(result2.errorMessage).toContain("Page number must be >= 1");
    console.log(`✓ Correctly rejected page=-1: ${result2.errorMessage}`);

    // Test 3: Out-of-range page number (page way beyond available data)
    console.log("\\nTest 3: page=999999 should return error (no more pages)");
    const result3 = await agentBay.list(
      { project: `list-test-${uniqueId}` },
      999999,
      2
    );
    expect(result3.success).toBe(false);
    expect(result3.errorMessage).toBeDefined();
    expect(result3.errorMessage).toContain("No more pages available");
    console.log(`✓ Correctly handled out-of-range page: ${result3.errorMessage}`);
  });

  test("should traverse all pages and verify pagination completeness", async () => {
    console.log("\\n=== Testing list() pagination completeness ===");

    const allSessionIds: string[] = [];
    let page = 1;
    const limit = 2;
    let totalCountFromApi = 0;

    console.log(`\\nTraversing all pages with limit=${limit}...`);

    while (true) {
      const result = await agentBay.list(
        { project: `list-test-${uniqueId}` },
        page,
        limit
      );

      // Verify each page request succeeds
      expect(result.success).toBe(true);
      expect(result.requestId).toBeDefined();

      // Collect session IDs
      allSessionIds.push(...result.sessionIds);
      totalCountFromApi = result.totalCount || 0;

      console.log(
        `Page ${page}: Found ${result.sessionIds.length} sessions, ` +
        `NextToken: ${result.nextToken ? "Yes" : "No"}, ` +
        `Total count: ${result.totalCount}`
      );

      // Verify page size constraint
      expect(result.sessionIds.length).toBeLessThanOrEqual(limit);

      // Check if we've reached the end
      if (!result.nextToken) {
        console.log(`\\n✓ Reached last page (page ${page})`);
        break;
      }

      page++;

      // Safety check to prevent infinite loop
      if (page > 100) {
        throw new Error("Pagination exceeded 100 pages, possible infinite loop");
      }
    }

    // Verify we collected at least our 3 test sessions
    expect(allSessionIds.length).toBeGreaterThanOrEqual(3);

    // Verify our test sessions are in the collected IDs
    const testSessionIds = testSessions.map((s) => s.sessionId);
    const foundCount = allSessionIds.filter((sid) =>
      testSessionIds.includes(sid)
    ).length;
    expect(foundCount).toBe(3);

    // Verify no duplicate session IDs
    const uniqueSessionIds = new Set(allSessionIds);
    expect(allSessionIds.length).toBe(uniqueSessionIds.size);

    console.log(`\\n✓ Pagination completeness verified:`);
    console.log(`  - Total pages traversed: ${page}`);
    console.log(`  - Total sessions collected: ${allSessionIds.length}`);
    console.log(`  - Total count from API: ${totalCountFromApi}`);
    console.log(`  - Found all 3 test sessions: Yes`);
    console.log(`  - No duplicates: Yes`);
  });

  test("should have consistent total_count across requests", async () => {
    console.log("\\n=== Testing list() total_count consistency ===");

    // Test 1: Verify total_count >= number of test sessions
    console.log("\\nTest 1: total_count should be >= 3 (our test sessions)");
    const result1 = await agentBay.list(
      { owner: `test-${uniqueId}` },
      undefined,
      10
    );

    expect(result1.success).toBe(true);
    expect(result1.totalCount).toBeGreaterThanOrEqual(3);
    console.log(`✓ total_count = ${result1.totalCount} (>= 3)`);

    // Test 2: Verify total_count remains consistent across multiple calls
    console.log("\\nTest 2: total_count should be consistent across multiple calls");
    const result2 = await agentBay.list(
      { owner: `test-${uniqueId}` },
      undefined,
      10
    );

    expect(result1.totalCount).toBe(result2.totalCount);
    console.log(`✓ total_count consistent: ${result1.totalCount} == ${result2.totalCount}`);

    // Test 3: Verify total_count matches actual sessions when collecting all pages
    console.log("\\nTest 3: total_count should match actual session count across all pages");

    const allSessionIds: string[] = [];
    let page = 1;
    const limit = 2;
    let expectedTotalCount = 0;

    while (true) {
      const result = await agentBay.list(
        { project: `list-test-${uniqueId}` },
        page,
        limit
      );

      if (!result.success) {
        break;
      }

      allSessionIds.push(...result.sessionIds);
      expectedTotalCount = result.totalCount || 0;

      if (!result.nextToken) {
        break;
      }

      page++;

      if (page > 100) {
        break;
      }
    }

    // The total_count should match the number of unique sessions collected
    const uniqueSessionIds = new Set(allSessionIds);
    expect(uniqueSessionIds.size).toBe(expectedTotalCount);

    console.log(
      `✓ total_count matches actual count: ` +
      `${expectedTotalCount} == ${uniqueSessionIds.size}`
    );

    // Test 4: Verify session_ids count <= total_count for each page
    console.log("\\nTest 4: session_ids count should be <= total_count on each page");
    const result = await agentBay.list(
      { owner: `test-${uniqueId}` },
      undefined,
      2
    );

    expect(result.sessionIds.length).toBeLessThanOrEqual(result.totalCount || 0);
    console.log(
      `✓ session_ids count (${result.sessionIds.length}) <= ` +
      `total_count (${result.totalCount})`
    );
  });
});

File: golang/tests/pkg/integration/agentbay_list_integration_test.go
package integration_test

import (
	"fmt"
	"math/rand"
	"testing"
	"time"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/agentbay/testutil"
)

func generateUniqueID() string {
	timestamp := time.Now().UnixNano() / 1000
	randomPart := rand.Intn(10000)
	return fmt.Sprintf("%d-%d", timestamp, randomPart)
}

// TestAgentBay_List_Integration tests the List API with end-to-end integration
func TestAgentBay_List_Integration(t *testing.T) {
	// Initialize AgentBay client
	apiKey := testutil.GetTestAPIKey(t)
	agentBayClient, err := agentbay.NewAgentBay(apiKey, nil)
	if err != nil {
		t.Fatalf("Error initializing AgentBay client: %v", err)
	}

	// Generate unique ID for this test run
	uniqueID := generateUniqueID()
	t.Logf("Using unique ID for test: %s", uniqueID)

	// Create multiple sessions with different labels for testing
	var testSessions []*agentbay.Session

	// Session 1: project=list-test, environment=dev
	t.Log("Creating session 1 with dev environment...")
	params1 := agentbay.NewCreateSessionParams().WithLabels(map[string]string{
		"project":     fmt.Sprintf("list-test-%s", uniqueID),
		"environment": "dev",
		"owner":       fmt.Sprintf("test-%s", uniqueID),
	})
	result1, err := agentBayClient.Create(params1)
	if err != nil {
		t.Fatalf("Error creating session 1: %v", err)
	}
	testSessions = append(testSessions, result1.Session)
	t.Logf("Session 1 created: %s", result1.Session.SessionID)
	t.Logf("Request ID: %s", result1.RequestID)

	// Session 2: project=list-test, environment=staging
	t.Log("Creating session 2 with staging environment...")
	params2 := agentbay.NewCreateSessionParams().WithLabels(map[string]string{
		"project":     fmt.Sprintf("list-test-%s", uniqueID),
		"environment": "staging",
		"owner":       fmt.Sprintf("test-%s", uniqueID),
	})
	result2, err := agentBayClient.Create(params2)
	if err != nil {
		t.Fatalf("Error creating session 2: %v", err)
	}
	testSessions = append(testSessions, result2.Session)
	t.Logf("Session 2 created: %s", result2.Session.SessionID)
	t.Logf("Request ID: %s", result2.RequestID)

	// Session 3: project=list-test, environment=prod
	t.Log("Creating session 3 with prod environment...")
	params3 := agentbay.NewCreateSessionParams().WithLabels(map[string]string{
		"project":     fmt.Sprintf("list-test-%s", uniqueID),
		"environment": "prod",
		"owner":       fmt.Sprintf("test-%s", uniqueID),
	})
	result3, err := agentBayClient.Create(params3)
	if err != nil {
		t.Fatalf("Error creating session 3: %v", err)
	}
	testSessions = append(testSessions, result3.Session)
	t.Logf("Session 3 created: %s", result3.Session.SessionID)
	t.Logf("Request ID: %s", result3.RequestID)

	// Wait a bit for sessions to be fully created
	time.Sleep(2 * time.Second)

	// Cleanup function
	defer func() {
		t.Log("Cleaning up: Deleting all test sessions...")
		for _, session := range testSessions {
			result, err := agentBayClient.Delete(session)
			if err != nil {
				t.Logf("Warning: Error deleting session %s: %v", session.SessionID, err)
			} else {
				t.Logf("Session %s deleted. Success: %v, Request ID: %s",
					session.SessionID, result.Success, result.RequestID)
			}
		}
	}()

	// Test 1: List all sessions without any label filter
	t.Run("ListAllSessions", func(t *testing.T) {
		t.Log("\n=== Testing List() without labels ===")

		result, err := agentBayClient.List(nil, nil, nil)
		if err != nil {
			t.Fatalf("Error listing all sessions: %v", err)
		}

		// Verify the result
		if result.RequestID == "" {
			t.Error("Request ID should be present")
		}
		if result.SessionIds == nil {
			t.Error("SessionIds list should not be nil")
		}

		t.Logf("Total sessions found: %d", result.TotalCount)
		t.Logf("Request ID: %s", result.RequestID)
	})

	// Test 2: List sessions with a single label filter
	t.Run("ListWithSingleLabel", func(t *testing.T) {
		t.Log("\n=== Testing List() with single label ===")

		result, err := agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			nil,
			nil,
		)
		if err != nil {
			t.Fatalf("Error listing sessions with single label: %v", err)
		}

		// Verify the result
		if result.RequestID == "" {
			t.Error("Request ID should be present")
		}
		if len(result.SessionIds) < 3 {
			t.Errorf("Should find at least 3 sessions, found %d", len(result.SessionIds))
		}

		// Verify all returned sessions have the expected label
		sessionIDs := make(map[string]bool)
		for _, s := range testSessions {
			sessionIDs[s.SessionID] = true
		}

		foundCount := 0
		for _, sessionID := range result.SessionIds {
			if sessionIDs[sessionID] {
				foundCount++
			}
		}

		if foundCount != 3 {
			t.Errorf("Should find exactly 3 test sessions, found %d", foundCount)
		}

		t.Logf("Found %d test sessions", foundCount)
		t.Logf("Total sessions with label: %d", len(result.SessionIds))
		t.Logf("Request ID: %s", result.RequestID)
	})

	// Test 3: List sessions with multiple label filters
	t.Run("ListWithMultipleLabels", func(t *testing.T) {
		t.Log("\n=== Testing List() with multiple labels ===")

		result, err := agentBayClient.List(
			map[string]string{
				"project":     fmt.Sprintf("list-test-%s", uniqueID),
				"environment": "dev",
			},
			nil,
			nil,
		)
		if err != nil {
			t.Fatalf("Error listing sessions with multiple labels: %v", err)
		}

		// Verify the result
		if result.RequestID == "" {
			t.Error("Request ID should be present")
		}
		if len(result.SessionIds) < 1 {
			t.Errorf("Should find at least 1 session, found %d", len(result.SessionIds))
		}

		// Verify the dev session is in the results
		devSessionID := testSessions[0].SessionID
		found := false
		for _, sessionID := range result.SessionIds {
			if sessionID == devSessionID {
				found = true
				break
			}
		}

		if !found {
			t.Error("Dev session should be in the results")
		}

		t.Logf("Found dev session: %v", found)
		t.Logf("Total matching sessions: %d", len(result.SessionIds))
		t.Logf("Request ID: %s", result.RequestID)
	})

	// Test 4: List sessions with pagination parameters
	t.Run("ListWithPagination", func(t *testing.T) {
		t.Log("\n=== Testing List() with pagination ===")

		// List first page with limit of 2
		page := 1
		limit := int32(2)
		resultPage1, err := agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			&page,
			&limit,
		)
		if err != nil {
			t.Fatalf("Error listing first page: %v", err)
		}

		// Verify first page
		if resultPage1.RequestID == "" {
			t.Error("Request ID should be present")
		}
		if len(resultPage1.SessionIds) > 2 {
			t.Errorf("First page should have at most 2 sessions, found %d", len(resultPage1.SessionIds))
		}

		t.Logf("Page 1 - Found %d sessions", len(resultPage1.SessionIds))
		t.Logf("Request ID: %s", resultPage1.RequestID)

		// If there are more results, test page 2
		if resultPage1.NextToken != "" {
			page2 := 2
			resultPage2, err := agentBayClient.List(
				map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
				&page2,
				&limit,
			)
			if err != nil {
				t.Fatalf("Error listing second page: %v", err)
			}

			if resultPage2.RequestID == "" {
				t.Error("Request ID should be present")
			}

			t.Logf("Page 2 - Found %d sessions", len(resultPage2.SessionIds))
			t.Logf("Request ID: %s", resultPage2.RequestID)
		}
	})

	// Test 5: List sessions with non-matching labels
	t.Run("ListWithNonMatchingLabels", func(t *testing.T) {
		t.Log("\n=== Testing List() with non-matching label ===")

		result, err := agentBayClient.List(
			map[string]string{
				"project":     fmt.Sprintf("list-test-%s", uniqueID),
				"environment": "nonexistent",
			},
			nil,
			nil,
		)
		if err != nil {
			t.Fatalf("Error listing sessions with non-matching label: %v", err)
		}

		// Verify the result
		if result.RequestID == "" {
			t.Error("Request ID should be present")
		}

		// Verify our test sessions are NOT in the results
		sessionIDs := make(map[string]bool)
		for _, s := range testSessions {
			sessionIDs[s.SessionID] = true
		}

		foundCount := 0
		for _, sessionID := range result.SessionIds {
			if sessionIDs[sessionID] {
				foundCount++
			}
		}

		if foundCount != 0 {
			t.Errorf("Should not find any test sessions, found %d", foundCount)
		}

		t.Logf("Correctly found %d test sessions (expected 0)", foundCount)
		t.Logf("Request ID: %s", result.RequestID)
	})

	// Test 6: Verify default limit
	t.Run("ListWithDefaultLimit", func(t *testing.T) {
		t.Log("\n=== Testing List() with default limit ===")

		result, err := agentBayClient.List(
			map[string]string{"owner": fmt.Sprintf("test-%s", uniqueID)},
			nil,
			nil,
		)
		if err != nil {
			t.Fatalf("Error listing sessions with default limit: %v", err)
		}

		// Verify the result
		if result.RequestID == "" {
			t.Error("Request ID should be present")
		}
		if result.MaxResults != 10 {
			t.Errorf("Default limit should be 10, got %d", result.MaxResults)
		}

		t.Logf("Max results: %d", result.MaxResults)
		t.Logf("Request ID: %s", result.RequestID)
	})

	// Test 7: Verify request_id is always present
	t.Run("ListRequestIDPresence", func(t *testing.T) {
		t.Log("\n=== Testing List() request_id presence ===")

		// Test 1: No labels
		result1, err := agentBayClient.List(nil, nil, nil)
		if err != nil {
			t.Fatalf("Error in test 1: %v", err)
		}
		if result1.RequestID == "" {
			t.Error("Test 1: Request ID should not be empty")
		}
		t.Logf("Test 1 Request ID: %s", result1.RequestID)

		// Test 2: With labels
		result2, err := agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			nil,
			nil,
		)
		if err != nil {
			t.Fatalf("Error in test 2: %v", err)
		}
		if result2.RequestID == "" {
			t.Error("Test 2: Request ID should not be empty")
		}
		t.Logf("Test 2 Request ID: %s", result2.RequestID)

		// Test 3: With pagination
		page := 1
		limit := int32(5)
		result3, err := agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			&page,
			&limit,
		)
		if err != nil {
			t.Fatalf("Error in test 3: %v", err)
		}
		if result3.RequestID == "" {
			t.Error("Test 3: Request ID should not be empty")
		}
		t.Logf("Test 3 Request ID: %s", result3.RequestID)
	})

	// Test 8: Error scenarios (invalid page numbers and out-of-range pages)
	t.Run("ListErrorScenarios", func(t *testing.T) {
		t.Log("\n=== Testing List() error scenarios ===")

		// Test 1: page=0
		t.Log("Test 1: Testing page=0")
		page0 := 0
		limit := int32(5)
		_, err := agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			&page0,
			&limit,
		)
		if err == nil {
			t.Error("page=0 should return an error")
		} else {
			t.Logf("page=0 correctly failed with error: %s", err.Error())
		}

		// Test 2: page=-1
		t.Log("Test 2: Testing page=-1")
		pageNeg := -1
		_, err = agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			&pageNeg,
			&limit,
		)
		if err == nil {
			t.Error("page=-1 should return an error")
		} else {
			t.Logf("page=-1 correctly failed with error: %s", err.Error())
		}

		// Test 3: page=999999 (out of range)
		t.Log("Test 3: Testing page=999999 (out of range)")
		pageHuge := 999999
		limitSmall := int32(2)
		_, err = agentBayClient.List(
			map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
			&pageHuge,
			&limitSmall,
		)
		if err == nil {
			t.Error("page=999999 should return an error (out of range)")
		} else {
			t.Logf("page=999999 correctly failed with error: %s", err.Error())
		}
	})

	// Test 9: Pagination completeness (traverse all pages until next_token is empty)
	t.Run("ListPaginationCompleteness", func(t *testing.T) {
		t.Log("\n=== Testing List() pagination completeness ===")

		allSessionIDs := make([]string, 0)
		page := 1
		limit := int32(2)
		maxIterations := 100

		for i := 0; i < maxIterations; i++ {
			result, err := agentBayClient.List(
				map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
				&page,
				&limit,
			)
			if err != nil {
				t.Fatalf("Error listing page %d: %v", page, err)
			}

			allSessionIDs = append(allSessionIDs, result.SessionIds...)

			t.Logf("Page %d: Found %d sessions", page, len(result.SessionIds))

			if result.NextToken == "" {
				t.Logf("Reached end of pagination at page %d", page)
				break
			}

			page++
		}

		// Verify we found all 3 test sessions
		testSessionIDs := make(map[string]bool)
		for _, session := range testSessions {
			testSessionIDs[session.SessionID] = true
		}

		foundCount := 0
		for _, sessionID := range allSessionIDs {
			if testSessionIDs[sessionID] {
				foundCount++
			}
		}

		if foundCount != 3 {
			t.Errorf("Should find all 3 test sessions, found %d", foundCount)
		}

		// Verify no duplicates
		uniqueIDs := make(map[string]bool)
		for _, sessionID := range allSessionIDs {
			if uniqueIDs[sessionID] {
				t.Errorf("Duplicate session ID found: %s", sessionID)
			}
			uniqueIDs[sessionID] = true
		}

		t.Logf("Total sessions collected: %d", len(allSessionIDs))
		t.Logf("Unique sessions: %d", len(uniqueIDs))
		t.Logf("Test sessions found: %d/3", foundCount)
	})

	// Test 10: total_count consistency
	t.Run("ListTotalCountConsistency", func(t *testing.T) {
		t.Log("\n=== Testing List() total_count consistency ===")

		// Test 1: total_count >= test sessions
		t.Log("Test 1: Verifying total_count >= test sessions")
		limit := int32(10)
		result1, err := agentBayClient.List(
			map[string]string{"owner": fmt.Sprintf("test-%s", uniqueID)},
			nil,
			&limit,
		)
		if err != nil {
			t.Fatalf("Error listing sessions: %v", err)
		}

		if result1.TotalCount < 3 {
			t.Errorf("total_count should be at least 3, got %d", result1.TotalCount)
		}
		t.Logf("total_count: %d (expected >= 3)", result1.TotalCount)

		// Test 2: total_count consistent across calls
		t.Log("Test 2: Verifying total_count consistency across calls")
		result2, err := agentBayClient.List(
			map[string]string{"owner": fmt.Sprintf("test-%s", uniqueID)},
			nil,
			&limit,
		)
		if err != nil {
			t.Fatalf("Error listing sessions: %v", err)
		}

		if result1.TotalCount != result2.TotalCount {
			t.Errorf("total_count should be consistent: first=%d, second=%d",
				result1.TotalCount, result2.TotalCount)
		}
		t.Logf("total_count consistency verified: %d", result2.TotalCount)

		// Test 3: total_count matches collected sessions count
		t.Log("Test 3: Verifying total_count matches actual session count")
		allSessionIDs := make([]string, 0)
		page := 1
		pageLimit := int32(2)

		for {
			result, err := agentBayClient.List(
				map[string]string{"project": fmt.Sprintf("list-test-%s", uniqueID)},
				&page,
				&pageLimit,
			)
			if err != nil {
				t.Fatalf("Error listing page %d: %v", page, err)
			}

			allSessionIDs = append(allSessionIDs, result.SessionIds...)

			if result.NextToken == "" {
				// Verify total_count matches
				uniqueIDs := make(map[string]bool)
				for _, sessionID := range allSessionIDs {
					uniqueIDs[sessionID] = true
				}

				if int32(len(uniqueIDs)) != result.TotalCount {
					t.Errorf("total_count mismatch: reported=%d, actual=%d",
						result.TotalCount, len(uniqueIDs))
				}
				t.Logf("total_count matches actual count: %d", result.TotalCount)
				break
			}

			page++
		}
	})
}
